src/audio-worklet/shim/spn-demo/app.mjs

```diff
@@ -0,0 +1,307 @@
+ */
+function setupRecording(recordBuffer) {
+  const recordButton = document.querySelector('#record');
+  const recordText = recordButton.querySelector('span');
+  const player = document.querySelector('#player');
+  const downloadButton = document.querySelector('#download');
+
+  async function prepareClip() {
+    // Get recording file URL
+    const wavUrl = createLinkFromAudioBuffer(recordBuffer, true);
+
+    // Calculate length
+    document.querySelector('#data-len').innerHTML =
+      Math.round(recordLength / recordBuffer.sampleRate * 100)/100;
+
+    // Set URL on player and download button
+    player.src = wavUrl;
+    downloadButton.src = wavUrl;
+    downloadButton.download = 'recording.wav';
+  }
+
+  recordButton.addEventListener('click', (e) => {
+    isRecording = !isRecording;
+
+    // When recording is paused, process clip
+    if (!isRecording) {
+      prepareClip();
+    }
+
+    recordText.innerHTML = isRecording ? 'Stop' : 'Start';
+  });
+}
+
+/**
+ * Sets up all visualizers.
+ * @param {AnalyserNode} liveAnalyser Live Analyser node to
+ * render live frequency graph from
+  */
+function setupVisualizers(liveAnalyser) {
+  setupLiveGainVis();
+  setupLiveAnalyserVis(liveAnalyser);
+  setupRecordingGainVis();
+
+  const visToggle = document.querySelector('#viz-toggle');
+  visToggle.addEventListener('click', (e) => {
+    visualizationEnabled = !visualizationEnabled;
+    visToggle.querySelector('span').innerHTML = visualizationEnabled ?
+            'Pause' :
+            'Play';
+  });
+}
+
+/**
+ * Script Processor Frequency Visualizer
+ */
+const setupLiveGainVis = () => {
+  const canvas = document.querySelector('#live-canvas');
+  const canvasContext = canvas.getContext('2d');
+
+  const width = canvas.width;
+  const height = canvas.height;
+
+  const drawStart = width-1;
+
+  const draw = () => {
+    requestAnimationFrame(draw);
+    if (!visualizationEnabled || !currData) return;
+
+    // Determine center and gain height
+    const centerY = ((1 - currDataGain) * height) / 2;
+    const gainHeight = currDataGain * height;
+
+    // Fill gain
+    canvasContext.fillStyle = 'black';
+    canvasContext.fillRect(drawStart, centerY, 1, gainHeight);
+
+    // Copy visualizer left
+    canvasContext.globalCompositeOperation = 'copy';
+    canvasContext.drawImage(canvas, -1, 0);
+    canvasContext.globalCompositeOperation = 'source-over';
+  };
+
+  draw();
+};
+
+/**
+ * Analyser Frequency Visualizer
+ * @param {AnalyserNode} analyserNode Analyser node to
+ * interpret frequency data from
+ */
+function setupLiveAnalyserVis(analyserNode) {
+  const canvas = document.querySelector('#analyzer-vis');
+  const canvasContext = canvas.getContext('2d');
+
+  const bufferLength = analyserNode.frequencyBinCount;
+  const dataArray = new Uint8Array(bufferLength);
+
+  const width = canvas.width;
+  const height = canvas.height;
+  const barWidth = width / bufferLength;
+
+  function draw() {
+    requestAnimationFrame(draw);
```

Comment:
By this way, you're putting multiple render functions into rAF. I think it's better to have 1 main render function and render everything there in the order you intended.